plugins/SlicerT/SlicerT.cpp

```diff
@@ -0,0 +1,548 @@
+/*
+ * SlicerT.cpp - simple slicer plugin
+ *
+ * Copyright (c) 2006-2008 Andreas Brandmaier <andy/at/brandmaier/dot/de>
+ * 
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+// TODO: add fft cache, maybe this is overkill but whatever
+// TODO: optimize waveform a LOT, mostly during playback
+
+#include "SlicerT.h"
+
+#include <fftw3.h>
+#include <QDomElement>
+
+#include "Engine.h"
+#include "Song.h"
+#include "InstrumentTrack.h"
+
+#include "PathUtil.h"
+#include "embed.h"
+#include "plugin_export.h"
+
+
+namespace lmms
+{
+
+extern "C"
+{
+Plugin::Descriptor PLUGIN_EXPORT slicert_plugin_descriptor =
+{
+	LMMS_STRINGIFY( PLUGINhandleAME ),
+	"SlicerT",
+	QT_TRANSLATE_NOOP( "PluginBrowser",
+				"Basic Slicer" ),
+	"Daniel Kauss Serna <daniel.kauss.serna/at/gmail.com>",
+	0x0100,
+	Plugin::Type::Instrument,
+	new PluginPixmapLoader( "logo" ),
+	nullptr,
+	nullptr,
+} ;
+} // end extern
+
+
+SlicerT::SlicerT(InstrumentTrack * instrumentTrack) : 
+	Instrument( instrumentTrack, &slicert_plugin_descriptor ),
+	m_noteThreshold(0.6f, 0.0f, 2.0f, 0.01f, this, tr( "Note threshold" ) ),
+	m_fadeOutFrames(0.0f, 0.0f, 8192.0f, 4.0f, this, tr("FadeOut")),
+	m_originalBPM(1, 1, 999, this, tr("Original bpm")),
+	m_originalSample()
+{}
+
+
+void SlicerT::playNote( NotePlayHandle * handle, sampleFrame * workingBuffer ) {
+	if (m_originalSample.frames() < 2048) {
+		return;
+	}
+	const float speedRatio = (float)m_originalBPM.value() / Engine::getSong()->getTempo() ;
+	const int noteIndex = handle->key() - 69;
+	const fpp_t frames = handle->framesLeftForCurrentPeriod();
+	const f_cnt_t offset = handle->noteOffset();
+	const int playedFrames = handle->totalFramesPlayed();
+
+	if (m_currentSpeedRatio != speedRatio) {
+		timeShiftSample();
+	}
+	const int totalFrames = m_timeShiftedSample.frames();
+
+	int sliceStart, sliceEnd;
+	if (noteIndex > m_slicePoints.size()-2 || noteIndex < 0) {
+		sliceStart = 0;
+		sliceEnd = totalFrames;
+	} else {
+		sliceStart = m_slicePoints[noteIndex] * speedRatio;
+		sliceEnd = m_slicePoints[noteIndex+1] * speedRatio;
+	}
+	
```

Comment:
There are tracing whitespace in some places. As an example, Line 95 above have a tab in it that doesn't show in the GitHub gui.

Reply:
Should be all gone now